### DISCUSSION TITLE
Agents in a state different from default state from the XML file

### DISCUSSION THREAD

**Discussion by daniele-baccega** [FLAMEGPU/FLAMEGPU2#1091] [Q&A] [answered]
Hi,
I tried to initialise an agent to a particular state (different from the default state) using the `<state>...</state>` tag from the XML file, but it does not work because I do not have that agent inside the simulation (I also created the state for the agent using the function `newState(...)`). Do you know why?

Thank you very much

**Comment by Robadob** [comment]
I've just looked at the IO [test case](https://github.com/FLAMEGPU/FLAMEGPU2/blob/master/tests/test_cases/io/test_io.cu#L119), this covers the behaviour you're having a problem with.

It has two agents:
* Agent `a`:
   * State: `default` (no bespoke states)
* Agent `b`:
   * State: `1`
   * State: `2`
   
It exports and reimports agents in the `a:default` and `b:2` states. I'm not aware of this test failing recently, so could you could please provide some greater detail so that I can help you find whether you've made a mistake or discovered an edge-case we're not testing for.

**Reply by daniele-baccega** [reply]
Inside the XML file I have something like this:
```
...
<xagent>
        <name>pedestrian</name>
        <state>susceptible</state>
        <x type="float">0.0</x>
        <y type="float">0.0</y>
        <z type="float">0.0</z>
        ...
</xagent>
...

```
And inside the main I declared:
```
AgentDescription pedestrian = model.newAgent("pedestrian");
pedestrian.newVariable<float>("x");
pedestrian.newVariable<float>("y");
pedestrian.newVariable<float>("z");
...
pedestrian.newState("susceptible");
...
```

When I run the simulation, I obtain the following error:
`terminate called after throwing an instance of 'flamegpu::exception::InvalidAgentState'
what():  /home/trustalert/Desktop/cottolengo-hospital-flame-gpu-2/build/_deps/flamegpu2-src/src/flamegpu/simulation/LoggingConfig.cu(31): State 'default' was not found within agent 'pedestrian' in the model description, in LoggingConfig::agent()`

**Reply by Robadob** [reply]
Thanks, I'll see if I can reproduce that on my end first thing tomorrow morning when I get to the office.

**Reply by Robadob** [reply]
Ah, reading your last comment more closely, your exception is due to `LoggingConfig` rather than state input/output files.

I would assume you've followed the [documentation](https://docs.flamegpu.com/guide/running-multiple-simulations/index.html#creating-a-logging-configuration). This isn't too clear, but you need to pass the state name as the second argument if you're using agent states (header for the function is [here](https://github.com/FLAMEGPU/FLAMEGPU2/blob/193fd8319bc32cc1332d4420c57a11ed7033de69/include/flamegpu/simulation/LoggingConfig.h#L114)).

I'll create a PR to improve that page of the documentation today.

We also already have an open issue to improve how default state is handled when using agent states, #907, it's possible that will be addressed in the next few months (although most likely will be September at the earliest as there's alot of annual leave over summer).

**Reply by Robadob** [reply]
I noticed that agent states are mentioned with regards to logging in the docs for non-batch runs ([here](https://docs.flamegpu.com/guide/running-a-simulation/collecting-data.html#configuring-data-to-be-logged)).

Regardless, I've created a PR to improve the other page: https://github.com/FLAMEGPU/FLAMEGPU2-docs/pull/155

**Reply by daniele-baccega** [reply]
I fixed the error relative to the log, but now I obtain this other one:
```
terminate called after throwing an instance of 'flamegpu::exception::InvalidAgentState'
  what():  /home/trustalert/Desktop/cottolengo-hospital-flame-gpu-2/build/_deps/flamegpu2-src/src/flamegpu/simulation/detail/CUDAAgent.cu(133): Agent 'default' does not use the default state, so the state must be passed explicitly, in CUDAAgent::setPopulationData()
```

**Reply by Robadob** [reply]
Similar to the logging issue when calling `CUDASimulation::setPopulationData()`, you will need to pass your state's name as the second argument (otherwise it defaults to `default`, header [here](https://github.com/FLAMEGPU/FLAMEGPU2/blob/master/include/flamegpu/simulation/CUDASimulation.h)).

**Reply by daniele-baccega** [reply]
But I do not call the function CUDASimulation::setPopulationData() because I'm creating and initializing the agents from the XML file specifying in the <state> tag the desired state

**Reply by Robadob** [reply]
Hmm, I'll investigate this further later this afternoon (I'm in and out of meetings for the next 2 hrs)

**Reply by Robadob** [reply, marked as answer]
Ah found the bug.

https://github.com/FLAMEGPU/FLAMEGPU2/blob/193fd8319bc32cc1332d4420c57a11ed7033de69/src/flamegpu/simulation/Simulation.cu#L199

Should be `setPopulationData(*agent.second, agent.first.second);`.

I'll create a quick PR, can merge it after CI passes (probably ~2 hours).

**Reply by Robadob** [reply]
https://github.com/FLAMEGPU/FLAMEGPU2/pull/1093

Fix has now been merged into `master`.

**Reply by daniele-baccega** [reply]
It works, thank you very much!
